lib/types.nix

```diff
@@ -352,6 +352,25 @@ rec {
       functor = (defaultFunctor name) // { wrapped = [ t1 t2 ]; };
     };
 
+    coercedTo = coercedType: coerceFunc: finalType: assert coercedType.getSubModules == null; mkOptionType rec {
+      name = "coercedTo";
+      description = "${finalType.description} or ${coercedType.description}";
+      check = x: finalType.check x || coercedType.check x;
+      merge = loc: defs:
+        let
+          coerceVal = val: if finalType.check val
+                           then val
+                           else let
+                                  coerced = coerceFunc val;
+                                in assert finalType.check coerced; coerced;
+
+        in finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
+      getSubOptions = finalType.getSubOptions;
+      getSubModules = finalType.getSubModules;
+      substSubModules = m: coercedTo coercedType coerceFunc (finalType.substSubModules m);
+      functor = (defaultFunctor name) // { wrapped = finalType; };
```

Comment:
This would not work if multiple options are declared with the same type.
Also, merging multiple `coerceTo` types or even a `coerceTo` with a `finalType` might not be easy, add a failing test case and a nicer error message.